File: 99.recover-binary-search-tree.cpp
/*
 * @lc app=leetcode id=99 lang=cpp
 *
 * [99] Recover Binary Search Tree
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // T: O(N), N := the number of nodes
    // S: O(1) (C/C++ tail recursion)
    void recoverTree(TreeNode* root) {
        n1 = nullptr;
        n2 = nullptr;
        pre = nullptr;

        traverse(root);
        int tmp = n1->val;
        n1->val = n2->val;
        n2->val = tmp;
    }

private:
    TreeNode *n1, *n2, *pre;

    // Inorder traversal that finds the problematic pair of nodes
    void traverse(TreeNode* node) {
        if (!node)  return;
        traverse(node->left);

        if (pre) {
            if (node->val < pre->val) {
                n2 = node;
                if (!n1) {
                    n1 = pre;
                }
            }
        }
        // if (!n1 && (!pre || node->val < pre->val)){
        //     n1 = pre;
        // }
        // if (n1 && node->val < pre->val){
        //     n2 = node;
        // }
        pre = node;

        traverse(node->right);
    }
};
// @lc code=end

File: 99.recover-binary-search-tree.py
#
# @lc app=leetcode id=99 lang=python3
#
# [99] Recover Binary Search Tree
#
from typing import *

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # T: O(n)
    # S: O(n) recursion stack
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        self.node1 = None
        self.node2 = None

        # The last visited node in inorder traversal. It is initialized to 
        # INT_MIN so that the logic in inorder traversal is simpler.
        self.pre = TreeNode(-2 ** 31)

        self.inorder(root)

        # Swap node values
        self.node1.val, self.node2.val = self.node2.val, self.node1.val

    # Remember, the result of the inorder traversal of a correct BST is sorted.
    # Thus, we can do an inorder traversal and find the only 2 node values that 
    # are greater than their next values.
    def inorder(self, node: Optional[TreeNode]):
        if not node:
            return
        self.inorder(node.left)
        
        # Example: original traversal result [6, 3, 4, 5, 2]
        #   - Check if the value previously visited is greater
        #   - If node1 is not yet found, set node1 to that previous node (val = 6)
        #   - If node1 is found, set node2 to the current node (val = 2)
        # if not self.node1 and node.val < self.pre.val:
        #     self.node1 = self.pre
        # if self.node1 and node.val < self.pre.val:
        #     self.node2 = node
        if node.val < self.pre.val:
            if not self.node1:
                self.node1 = self.pre
            self.node2 = node

        # The visited node
        self.pre = node

        self.inorder(node.right)
        
# @lc code=end

